### PR TITLE
Add a "bump allocator", a small and simple allocator for Wasm

### PR DESCRIPTION
**Pull request by Bargsteen**
## Purpose

This PR implements a new allocator for our smart contracts. It is simple and has a small code size. It uses more memory than a linked-list allocator such as wee_alloc, but for many contracts, this simplicity is sufficient and better.

Something I discovered while building this, is that the Rust compiler sets the initial number of pages in wasm modules to 17, which seems like quite a lot for smart contracts. Each page is 64KiB, so the initial memory is always ~1.1 MiB.

Closes https://github.com/Concordium/concordium-rust-smart-contracts/issues/379

## TODO
- The templates do not compile since they use the published version of concordium-std, so we need to publish it first, then merge this PR. Business as usual, but still a bit annoying and error-prone.

## Changes

- Implement a bump allocator
- Add a `bump_alloc` feature to `concordium_std`
- Add tests via the contract `bump-alloc-tests`
- Remove the `wee_alloc` feature **(breaking)**
- Change all contracts that use `wee_alloc` to use `bump_alloc`
- Update the `concordium-std` library documentation to reflect the new reality
- Bump MSRV to `1.72` in the testing lib to fix an issue with the indirect dependency `ahash` which requires MSRV `1.72`.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-std/src/bump_alloc.rs
Comment (limemloh) [inline concordium-std/src/bump_alloc.rs:183]: **Suggestion:** I did not test this, but I think we could avoid branching here, maybe the compiler already does this already internally.
```suggestion
    let rounding = r > 0 as usize;
    d + rounding
```

File: concordium-std/src/bump_alloc.rs
Comment (limemloh) [inline concordium-std/src/bump_alloc.rs:105]: **Question:** Why is this done as part of `alloc` and not in `BumbAllocator::new`?

**Review by limemloh — APPROVED** [review]
Look great, very simple :+1:

File: concordium-std/src/bump_alloc.rs
Comment (DOBEN) [inline concordium-std/src/bump_alloc.rs:5]:
```suggestion
//! the heap to hand out. The `next` pointer is incremented by the size of
```

File: concordium-std/src/bump_alloc.rs
Comment (DOBEN) [inline concordium-std/src/bump_alloc.rs:8]:
```suggestion
//! memory leaks are perfectly fine and even preferable since the need for
```

File: concordium-std/src/bump_alloc.rs
Comment (DOBEN) [inline concordium-std/src/bump_alloc.rs:52]: Suggestion: In case that value changes

```suggestion
    /// The number of active allocations. Used for resetting `next` to `MIN_PTR_ADDR` if
```

File: concordium-std/src/bump_alloc.rs
Comment (DOBEN) [inline concordium-std/src/bump_alloc.rs:77]: Suggestion: You could mentioned that the function returns the previous `PageCount`.

**Review by DOBEN — APPROVED** [review]
Tried it out and well done.

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs:105]: I'll document that, thanks for pointing it out.
The reason is that the `new` must be a `const` method so we can declare a static with it.

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs:26]: Ah, that makes sense. I didn't realize that the null pointer is the same as `0usize`. I thought it was some separate sentinel value or something like it.

File: concordium-std/src/bump_alloc.rs
Comment (abizjak) [reply concordium-std/src/bump_alloc.rs:191]: Final comment, since this is only used when rhs = PAGE_SIZE, I think in that special case this would be
```
lhs >> 16 + (lhs & 0xffff != 0 as usize)
```

File: concordium-std/src/bump_alloc.rs
Comment (abizjak) [inline concordium-std/src/bump_alloc.rs:76]: Are we sure we are not breaking some invariants with implementing this?

I'm guessing not

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs]: Fixed before you wrote this ;)

File: concordium-std/src/bump_alloc.rs
Comment (abizjak) [reply concordium-std/src/bump_alloc.rs:96]: No, let's do this.

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs]: I tried out your proposed solution Ales, but it seemed off by one in a few cases. It also isn't the easiest thing to read, and the compiler may end up doing something similar on its own.
But as you pointed out, the `rhs` is constant, so I created a more bespoke function for calculating it.

Feel free to add the bitwise arithmetic if you find it essential.

Comment (abizjak) [reply concordium-std/src/bump_alloc.rs]: Hmm, did you use what I intended rather than what I wrote (the parenthesis)
```
fn f1(xs: usize) -> usize {
    (xs >> 16) + ((xs & 0xffff != 0) as usize)
}
```
This program
```
fn f1(xs: usize) -> usize {
    (xs >> 16) + ((xs & 0xffff != 0) as usize)
}

const N: usize = 1 << 16;

fn f2(xs: usize) -> usize {
    let a = xs / N;
    if xs % N > 0 {
        a + 1
    } else {
        a
    }
}

fn main() {
    for x in 0..usize::MAX {
        assert_eq!(f1(x), f2(x), "Failed for {x}");
    }
}
```
Does not panic.

Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs]: Ah, turns out I had `PAGE_SIZE` set to `65535` in the Rust playground instead of `65536`, which caused the discrepancy. I've used your suggested approach now ;)

Comment (abizjak) [reply concordium-std/src/bump_alloc.rs]: :)

File: concordium-std/src/bump_alloc.rs
Comment (abizjak) [inline concordium-std/src/bump_alloc.rs:105]: Could you document what this field is for, and why it is needed?

File: concordium-std/src/bump_alloc.rs
Comment (abizjak) [inline concordium-std/src/bump_alloc.rs:132]: Would it be possible to rename this to `memory_size` or `linear_memory_size` since I got confused for a bit thinking this is the size of the heap.

File: concordium-std/src/lib.rs
Comment (abizjak) [inline concordium-std/src/lib.rs:134]: Perhaps qualify here that for most smart contracts such as tokens this is the right tradeoff.

For very complex ones perhaps it's not the right choice.

**Review by abizjak — APPROVED** [review]
Beautiful.

File: examples/bump-alloc-tests/tests/tests.rs
Comment (abizjak) [inline examples/bump-alloc-tests/tests/tests.rs:66]: Could you please add a test where you 
- have some fixed static string
- allocate some non-trivial amount, say 500kB of data
- make a function call with non-trivial parameters, and that function does something with the parameters
- recurse
- do something more in the function with the static string and parameters or some locals in the function defined before the recursive call.

I want us to have a test to make sure that we are not overwriting the stack or the static data in the allocator.

File: concordium-std/CHANGELOG.md
Comment (Bargsteen) [reply concordium-std/CHANGELOG.md:9]: Ah, classic merge error. Good catch ;)

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs:105]: It is documented on the type. But I also added a line here.

File: concordium-std/src/bump_alloc.rs
Comment (Bargsteen) [reply concordium-std/src/bump_alloc.rs:132]: Yep. Renamed to `memory_size` to match the actual name in wasm. I also renamed `grow_memory` to `memory_grow` for the same reason.

File: examples/bump-alloc-tests/tests/tests.rs
Comment (Bargsteen) [reply examples/bump-alloc-tests/tests/tests.rs:66]: Yes, I've added that now. Let me know if you deem it sufficient ;)

**Comment by abizjak** [comment]
@Bargsteen I suppose this is just waiting for us to release concordium-std/testing library before it can be merged?

**Comment by Bargsteen** [comment]
> @Bargsteen I suppose this is just waiting for us to release concordium-std/testing library before it can be merged?

Yep :-)

**Pull request merged**
